File: dsa3/day-36-linkedlist/ReverseLinkedList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class ReverseLinkedList:
    """
    This class provides a method to reverse a singly linked list in-place and in one pass.
    """

    # @param A : head node of the linked list
    # @return the head node of the reversed linked list
    def reverseList(self, A):
        # If the list has only one node, no reversal is needed
        if A is None or A.next is None:
            return A

        prev = None  # Tracks the previous node
        current = A  # Starts at the head of the list
        
        # Traverse the list and reverse links one node at a time
        while current:
            temp = current.next  # Store the next node
            current.next = prev  # Reverse the current node's pointer
            prev = current       # Move prev to the current node
            current = temp        # Move current to the next node
        
        # Return the new head of the reversed list
        return prev

    """
    Example Intuition:
    ------------------
    Consider a list: 1 -> 2 -> 3 -> None
    Step-by-step process of reversing the list:

    Initial state:
    prev = None, current = 1 -> 2 -> 3 -> None

    Step 1:
    temp = 2 -> 3 -> None  (store next node)
    current.next = None    (reverse current's pointer to prev)
    prev = 1 -> None       (move prev forward)
    current = 2 -> 3 -> None (move current forward)

    Step 2:
    temp = 3 -> None
    current.next = 1 -> None
    prev = 2 -> 1 -> None
    current = 3 -> None

    Step 3:
    temp = None
    current.next = 2 -> 1 -> None
    prev = 3 -> 2 -> 1 -> None
    current = None (end of list)

    Final reversed list: 3 -> 2 -> 1 -> None

    This approach ensures an in-place reversal in one traversal.
    """

# Example usage and explanation
def print_list(head):
    """Utility function to print the linked list."""
    result = []
    while head:
        result.append(str(head.val))
        head = head.next
    return " -> ".join(result) + " -> None"

# Example 1: Original list 1 -> 2 -> 3 -> None
head1 = ListNode(1)
head1.next = ListNode(2)
head1.next.next = ListNode(3)

reverse_linked_list = ReverseLinkedList()

print("Original List: ", print_list(head1))
reversed_head1 = reverse_linked_list.reverseList(head1)
print("Reversed List: ", print_list(reversed_head1))

# Example 2: Single node list 1 -> None
head2 = ListNode(1)
print("\nOriginal List: ", print_list(head2))
reversed_head2 = reverse_linked_list.reverseList(head2)
print("Reversed List: ", print_list(reversed_head2))

# Example 3: Empty list
head3 = None
print("\nOriginal List: None")
reversed_head3 = reverse_linked_list.reverseList(head3)
print("Reversed List: None")

# Example 4: Longer list 10 -> 20 -> 30 -> 40 -> None
head4 = ListNode(10)
head4.next = ListNode(20)
head4.next.next = ListNode(30)
head4.next.next.next = ListNode(40)

print("\nOriginal List: ", print_list(head4))
reversed_head4 = reverse_linked_list.reverseList(head4)
print("Reversed List: ", print_list(reversed_head4))

File: dsa3/day-36-linkedlist/ReverseLinkedList.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     public int val;
 *     public ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
public class ReverseLinkedList {

    /**
     * Reverses a singly linked list in-place and in one pass.
     * @param head The head node of the singly linked list.
     * @return The head node of the reversed linked list.
     */
    public ListNode reverseList(ListNode head) {
        ListNode prev = null, current = null;
        if (head == null || head.next == null) {
            return head; // If the list is empty or has one node, no need to reverse.
        }
        current = head;
        while (head != null) {
            ListNode t = head.next; // Store the next node.
            head.next = prev;       // Reverse the current node's pointer.
            prev = head;            // Move prev forward to the current node.
            head = t;               // Move head forward to the next node.
        }
        return prev; // Return the new head of the reversed list.
    }

    public static void main(String[] args) {
        ReverseLinkedList solution = new ReverseLinkedList();

        // Test case 1: 1 -> 2 -> 3 -> null
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(3);
        System.out.println("Original List: " + printList(head1));
        ListNode reversed1 = solution.reverseList(head1);
        System.out.println("Reversed List: " + printList(reversed1));

        // Test case 2: Empty list (null)
        ListNode head2 = null;
        System.out.println("Original List: " + printList(head2));
        ListNode reversed2 = solution.reverseList(head2);
        System.out.println("Reversed List: " + printList(reversed2));

        // Test case 3: Single node (1 -> null)
        ListNode head3 = new ListNode(1);
        System.out.println("Original List: " + printList(head3));
        ListNode reversed3 = solution.reverseList(head3);
        System.out.println("Reversed List: " + printList(reversed3));

        // Test case 4: 5 -> 10 -> 15 -> 20 -> null
        ListNode head4 = new ListNode(5);
        head4.next = new ListNode(10);
        head4.next.next = new ListNode(15);
        head4.next.next.next = new ListNode(20);
        System.out.println("Original List: " + printList(head4));
        ListNode reversed4 = solution.reverseList(head4);
        System.out.println("Reversed List: " + printList(reversed4));
    }

    /**
     * Utility method to print a linked list.
     * @param head The head node of the linked list.
     * @return A string representation of the linked list.
     */
    public static String printList(ListNode head) {
        StringBuilder sb = new StringBuilder();
        while (head != null) {
            sb.append(head.val).append(" -> ");
            head = head.next;
        }
        sb.append("null");
        return sb.toString();
    }
}

/**
 * Definition for the ListNode class.
 */
class ListNode {
    public int val;
    public ListNode next;
    ListNode(int x) { 
        val = x; 
        next = null; 
    }
}

/**
 * Example Explanation:
 *
 * Let's take an example linked list: 1 -> 2 -> 3 -> 4 -> null
 *
 * Step-by-step reversal process:
 *
 * 1. Initial State:
 *    prev = null, current = A (head pointing to 1), A = 1
 *    List: 1 -> 2 -> 3 -> 4 -> null
 *
 * 2. First Iteration:
 *    t = A.next (2), A.next = prev (null)
 *    prev = A (1), A = t (2)
 *    List: 1 -> null, 2 -> 3 -> 4 -> null
 *
 * 3. Second Iteration:
 *    t = A.next (3), A.next = prev (1)
 *    prev = A (2), A = t (3)
 *    List: 2 -> 1 -> null, 3 -> 4 -> null
 *
 * 4. Third Iteration:
 *    t = A.next (4), A.next = prev (2)
 *    prev = A (3), A = t (4)
 *    List: 3 -> 2 -> 1 -> null, 4 -> null
 *
 * 5. Fourth Iteration:
 *    t = A.next (null), A.next = prev (3)
 *    prev = A (4), A = t (null)
 *    List: 4 -> 3 -> 2 -> 1 -> null
 *
 * 6. End of Loop:
 *    A becomes null, prev points to the new head of the reversed list (4).
 *
 * Final Reversed List: 4 -> 3 -> 2 -> 1 -> null
 *
 * This ensures the linked list is reversed in-place in O(n) time and O(1) space.
 */
